adxl345/adxl345.go

```diff
@@ -0,0 +1,230 @@
+// The bus and chip parameters define the SPI bus and chip select to use.
+// The SPI s is configured.
+// The device is turned on.
+// The device is verified to be an ADXL345.
+func New(p spi.Port, o *Opts) (*Dev, error) {
+	// Convert the spi.Port into a spi.Conn so it can be used for communication.
+	c, err := p.Connect(SpiFrequency, SpiMode, SpiBits)
+	if err != nil {
+		return nil, err
+	}
+	d := &Dev{
+		name: "ADXL345",
+		s:    c,
+	}
+	if o.TurnOnOnStart {
+		err = d.TurnOn()
+		if err != nil {
+			return nil, err
+		}
+	}
+	if o.Sensitivity != S2G { // default
+		err = d.setSensitivity(o.Sensitivity)
+		if err != nil {
+			return nil, err
+		}
+	}
+	// Verify that the device is an ADXL345.
+	rx, _ := d.ReadRaw(DeviceID)
+	if rx[1] != o.ExpectedDeviceID {
+		return nil, fmt.Errorf("wrong device connected should be an adxl345  should be\"%#x\" rx0=\"%#x\" rx1=\"%#x\"", o.ExpectedDeviceID, rx[0], rx[1])
+	}
+	return d, nil
+}
+
+// SetSensitivity sets the sensitivity of the ADXL345.
+// The sensitivity parameter should be one of 2, 4, 8, or 16, representing ±2g, ±4g, ±8g, or ±16g respectively.
+func (d *Dev) setSensitivity(sensitivity Sensitivity) error {
+	switch sensitivity {
+	case S2G, S4G, S8G, S16G:
+		// Write to the DataFormat register
+		d.sensitivity = sensitivity
+		return d.Write(DataFormat, byte(sensitivity))
+	default:
+		return fmt.Errorf("invalid sensitivity: %d. Valid values are 2, 4, 8, 16", sensitivity)
+	}
+}
+
+// TurnOn turns on the measurement mode of the ADXL345.
+// This is required before reading data from the device.
+func (d *Dev) TurnOn() error {
+	return d.Write(PowerCtl, 0x08)
+}
+
+// TurnOff turns off the measurement mode of the ADXL345.
+func (d *Dev) TurnOff() error {
+	return d.Write(PowerCtl, 0x00)
+}
+
+// Update reads the acceleration values from the ADXL345.
+// By reading the acceleration the 3 axes acceleration values.
+// This is a simple synchronous implementation.
+func (d *Dev) Update() Acceleration {
+	return Acceleration{
+		X: d.ReadAndCombine(DataX0, DataX1),
+		Y: d.ReadAndCombine(DataY0, DataY1),
+		Z: d.ReadAndCombine(DataZ0, DataZ1),
+	}
+}
+
+// ReadAndCombine combines two registers to form a 16-bit value.
+// The ADXL345 uses two 8-bit registers to store the output data for each axis.
+// X := d.ReadAndCombine(DataX0, DataX1) where:
+// `DataX0` is the address of the lower byte (LSB, least significant byte)
+// `DataX1` is the address of the upper byte (MSB, most significant byte)
+// The ADXL345 combines both registers to deliver 16-bit output for each acceleration axis.
+// A similar approach is used for the Y and Z axes. This technique provides higher precision in the measurements.
+func (d *Dev) ReadAndCombine(reg1, reg2 byte) int16 {
+	low, _ := d.Read(reg1)
+	high, _ := d.Read(reg2)
+	return int16(uint16(high)<<8) | int16(low)
+}
+
+// Read reads a 16-bit value from the specified register address.
+func (d *Dev) Read(regAddress byte) (int16, error) {
```

Comment:
I'd recommend to use https://pkg.go.dev/periph.io/x/conn/v3@v3.7.0/mmr#Dev16.ReadUint16 instead and convert back to a Int16.

Reply:
@maruel Sorry, I m quite new to "periph" and i do not understand

```go
func (d *Dev) Read(regAddress byte) (int16, error) {
	tx := []byte{regAddress | 0x80, 0x00}
	rx := make([]byte, len(tx))
	err := d.s.Tx(tx, rx)
	if err != nil {
		return 0, err
	}
	return int16(binary.LittleEndian.Uint16(rx)), nil
}``` 
this conversion is equivalent to `return int16(uint16(rx[0]) | uint16(rx[1])<<8)`  i feel that using Dev16.ReadUint16 won't be more efficient.

Reply:
It's okay. The point is to reduce the amount of code as much as possible but it's not a big deal.